### ISSUE TITLE
API improvements

### ISSUE THREAD

**Issue by hadley** [hadley/httr#351]
* Need to more cleaning define the functions used for constructing requests from those that inspect the response. (`response_type()`, `response_status()` etc).

* Go back to `content_raw()`, `content_text()`, `content_parsed()`, and `content_auto()`. `content()` is too fragile to build on.

**Comment by hadley** [comment]
(Also note that I probably won't work on this for a while since I have higher priority stuff that I should be working, but this might happen next time I have a couple of days downtime and a good internet connection)

**Comment by jeroenooms** [comment]
It has been on my to do list to experiment with async libcurl bindings in order to support many concurrent requests. Something along the lines of https://github.com/jeroenooms/curl/issues/51 but perhaps a bit simpler, like [promises](http://www.html5rocks.com/en/tutorials/es6/promises/) in JavaScript.
 
```r
# New scheduler
pool <- multi_handle()

# Add requests with callback, returns immediately
req("http://url.com") %>% schedule(pool) %>% then(function(req){
  print("Done: ", req$status)
})

# This would block until all request are finished
complete(pool)
```

@jcheng5 and me briefly chatted about using the R event loop rather than something like `complete` and never block. But this might not be reliable on windows. 

Only vaguely related to the above :)

**Comment by craigcitro** [comment]
More to say later, but one quick thought to throw out there:

I do think it'd be a bummer to no longer have a simple `GET('http://some.url/')` call for interactive use. There's a fine line to walk here -- I much prefer the explicit composability described above, especially when it comes to testing, but it's often useful to be able to quickly grab a page to use for examples or while experimenting.

I'm not sure what the right answer is here -- I chatted about this with @hadley who was also on the fence. What do other folks think?

**Comment by jennybc** [comment]
> Should be a new package, httr2?

It feels like that would be easier on everyone. Maybe you should name first packages `foo0`, so you can just drop the zero on the second pass :)

> There's no consistent naming scheme that indicates whether a function works with a request or a response.
> ... large request functions that have many arguments passed in `...`

This rings true. I have wondered whether it's best to build the pieces first (e.g. the URL), then pass to the verb. Or just do everything at once and write a monster `VERB()` call. The pipe-based design clears some of that up. I have pieces of the request that are conditionally included (e.g. token if URL matches some regex, progress if interactive). How do you express that? BTW progress isn't mentioned in the request API notes.

> I do think it'd be a bummer to no longer have a simple GET('http://some.url/') call for interactive use.

I agree. And maybe not just for interactive use. When you're reading API docs, it's always in terms of `GET`, `POST`, etc. so it is nice to have those same verbs.

**Comment by hadley** [comment]
@jennybc my main worry about also having a simpler `GET()`, `POST()` etc set of functions is that it's going to be `qplot()` all over again - yes, it's easier to get started with, and it would be easier to copy examples from the web, but as soon as you want to do anything complicated, you'll want to switch to the full system.

(And progess would be something like `req(url) %>% req_show_progress() %>% req_fetch()`)

**Comment by jcheng5** [comment]
What if `req_fetch` can take a string? (Or really, any of the functions that operate on a request could take a string instead.)

**Comment by hadley** [comment]
@jcheng5  you mean like this example `req_fetch("http://url.com/")` 😛 

You could certainly start any chain with a string instead of a request, but I think you'd still need to explicitly send the request with `req_fetch()` etc.

**Comment by jennybc** [comment]
Will the design of `.httr-oauth` stay the same?

I know you are generally anti-`options()` but what about allowing us to set user agent package-wide with an option?

**Comment by jcheng5** [comment]
@hadley Reading comprehension FTW

@craigcitro If `req_fetch(url)` works, do you still need `GET(url)`? Or you're thinking that `GET` would take options as parameters?

**Comment by hadley** [comment]
@jennybc Yes, `.httr-oauth` is likely to stay the same (and you'd be able to use both httr and httr2 with the same file)

I'd really prefer not to use options, but I think it won't be as important since with the pipeable syntax, you'd just create a basic prototype request that every function in your package would customise.

**Comment by jeroenooms** [comment]
Wrote a first working version of fully async requests in the ‘curl’ package. See the `?multi` help page and examples. Maybe this helps rethinking the `httr` api?

**Comment by jennybc** [comment]
Re "Performing the request", it would be nice if the Request API offered a way to specify the behaviour available currently via `RETRY()`.

**Comment by hadley** [comment]
See also notes from @hrbrmstr at #396

**Comment by hadley** [comment]
Also need to break up OAuth apis into much smaller pieces as in #432

**Comment by hadley** [comment]
Need API to retrieve all requests performed when a sequence of requests are performed automatically (e.g. during redirects, #445)

**Comment by jennybc** [comment]
Current logic re: adding token cache file to `.Rbuildignore` and `.gitignore` only looks at directory of cache file. usethis now has a notion of a project. Would be good to use something similar to identify if there's a relevant `.Rbuildignore` or `.gitignore` in some parent directory of the cache file.
